File: python/102/patterns/decorator1.py
from abc import ABC, abstractmethod

# Interfejs bazowy
class INapoj(ABC):
    @abstractmethod
    def pobierz_opis(self) -> str:
        pass

    @abstractmethod
    def pobierz_koszt(self) -> float:
        pass


# Konkretny komponent
class Kawa(INapoj):
    def pobierz_opis(self) -> str:
        return "Kawa"

    def pobierz_koszt(self) -> float:
        return 5.00


# Dekorator bazowy
class NapojDekorator(INapoj):
    def __init__(self, napoj: INapoj):
        self._napoj = napoj

    def pobierz_opis(self) -> str:
        return self._napoj.pobierz_opis()

    def pobierz_koszt(self) -> float:
        return self._napoj.pobierz_koszt()


# Konkretny dekorator: Mleko
class Mleko(NapojDekorator):
    def pobierz_opis(self) -> str:
        return f"{self._napoj.pobierz_opis()}, Mleko"

    def pobierz_koszt(self) -> float:
        return self._napoj.pobierz_koszt() + 1.50


# Konkretny dekorator: Cukier
class Cukier(NapojDekorator):
    def pobierz_opis(self) -> str:
        return f"{self._napoj.pobierz_opis()}, Cukier"

    def pobierz_koszt(self) -> float:
        return self._napoj.pobierz_koszt() + 0.50


# Program
def main():
    kawa = Kawa()
    print(f"{kawa.pobierz_opis()} - {kawa.pobierz_koszt()} zł")

    kawa = Mleko(kawa)
    print(f"{kawa.pobierz_opis()} - {kawa.pobierz_koszt()} zł")

    kawa = Cukier(kawa)
    print(f"{kawa.pobierz_opis()} - {kawa.pobierz_koszt()} zł")


if __name__ == "__main__":
    main()

File: c#-basics/patterns/decorator1.cs
namespace ConsoleApp18
{
    using System;

    // Interfejs bazowy
    public interface INapoj
    {
        string PobierzOpis();
        double PobierzKoszt();
    }

    // Konkretny komponent
    public class Kawa : INapoj
    {
        public string PobierzOpis() => "Kawa";
        public double PobierzKoszt() => 5.00;
    }

    // Dekorator bazowy
    public abstract class NapojDekorator : INapoj
    {
        protected INapoj _napoj;

        public NapojDekorator(INapoj napoj)
        {
            _napoj = napoj;
        }

        public virtual string PobierzOpis() => _napoj.PobierzOpis();
        public virtual double PobierzKoszt() => _napoj.PobierzKoszt();
    }

    // Konkretny dekorator: Mleko
    public class Mleko : NapojDekorator
    {
        public Mleko(INapoj napoj) : base(napoj) { }

        public override string PobierzOpis() => $"{_napoj.PobierzOpis()}, Mleko";
        public override double PobierzKoszt() => _napoj.PobierzKoszt() + 1.50;
    }

    // Konkretny dekorator: Cukier
    public class Cukier : NapojDekorator
    {
        public Cukier(INapoj napoj) : base(napoj) { }

        public override string PobierzOpis() => $"{_napoj.PobierzOpis()}, Cukier";
        public override double PobierzKoszt() => _napoj.PobierzKoszt() + 0.50;
    }

    // Program
    class Program
    {
        static void Main(string[] args)
        {
            INapoj kawa = new Kawa();
            Console.WriteLine($"{kawa.PobierzOpis()} - {kawa.PobierzKoszt()} zł");

            INapoj kawaZMlekiem = new Mleko(kawa);
            Console.WriteLine($"{kawaZMlekiem.PobierzOpis()} - {kawaZMlekiem.PobierzKoszt()} zł");

            INapoj kawaZMlekiemICukrem = new Cukier(kawaZMlekiem);
            Console.WriteLine($"{kawaZMlekiemICukrem.PobierzOpis()} - {kawaZMlekiemICukrem.PobierzKoszt()} zł");
        }
    }
}
